bundles/org.openhab.core.karaf/src/main/java/org/openhab/core/karaf/internal/FeatureInstaller.java

```diff
@@ -451,17 +439,17 @@ private Set<String> getAllFeatureNamesWithPrefix(String prefix) {
                     .collect(Collectors.toSet());
         } catch (Exception e) {
             logger.error("Failed retrieving features: {}", e.getMessage(), debugException(e));
-            return Collections.emptySet();
+            return Set.of();
         }
     }
 
-    private synchronized void installFeatures(Set<String> addons) {
+    private void installFeatures(Set<String> addons) {
         try {
             if (logger.isDebugEnabled()) {
-                logger.debug("Installing '{}'", addons.stream().collect(Collectors.joining(", ")));
+                logger.debug("Installing '{}'", String.join(", ", addons));
             }
-            featuresService.installFeatures(addons,
-                    EnumSet.of(FeaturesService.Option.Upgrade, FeaturesService.Option.NoFailOnFeatureNotFound));
+            featuresService.installFeatures(addons, EnumSet.of(FeaturesService.Option.NoAutoRefreshBundles,
```

Comment:
@wborn Do you know if this might cause issues? The reason I added it is that some add-ons cause `org.openhab.core` to refresh if it is not set. As a result the `EventPublisher` is unloaded and this service is stopped. If a lot of add-ons are installed at the same time, this will result in an `InterruptedException` in `FeaturesService.installFeatures` which in turn shows up as ERROR in the log. At the next start everything is fixed.

It would be great if there is a documentation for the `FeaturesService.Option` (at least I didn't find it). Do you know of such a document?

Reply:
I don't think there is much documentation for the `FeaturesService` besides JavaDocs and the code itself. So if that doesn't help, mailing questions on the Karaf mailing list might help.

Maybe you can figure out what causes the bundle refresh in the first place and prevent it from happening?

The [Karaf docs](https://karaf.apache.org/manual/latest/#_feature_and_resolver) describes the reasons for automatically refreshing bundles:

> For instance, a bundle has to be refreshed:
> 
> 1. when bundle A uses an import package, and bundle B is installed providing a new version of the package (matching bundle A import version range). Then bundle A has to be refreshed to use the new version of the package.
> 2. when bundle A uses an optional import package, and bundle B is installed providing this package. Then, bundle A has to be refreshed to actually use the package.
> 3. when bundle A uses a package provided by bundle B, then, bundle A will be refreshed as well. This is kind of "cascading" refresh.

Reply:
> As a result the `EventPublisher` is unloaded and this service is stopped

Would the feature installation succeed when refreshing occurs while the `EventPublisher` has reference cardinaltiy optional? If that works, it could continue and for instance post any events whenever the `EventPublisher` is set again.

Reply:
I don't think that makes a difference. If the `EventPublisher` is removed, this is most likely because the `org.openhab.core` bundle is refreshed. This will automatically refresh the `org.openhab.core.karaf` bundle and the service is de-activated anyway. I now removed immediate refreshing during installation and refresh after all config updates have been processed (but only if the add-on configuration changed).